File: javascript/src/find-w.js
const { Helper } = require('./helper.js');
const { W } = require('./w.js');

class FindW {
    constructor(fn) {
        this.helper = new Helper();
        this.fn = fn;
    }

    find(ot) {
        const data = this.helper.l(this.fn);
        const wData = this.p(data);
        const w = this.helper.find(wData, ot);
        return w.r();
    }

    p(data) {
        return this.helper.p(data, this.startsWithDayOfMonth, this.newW);
    }

    startsWithDayOfMonth(s) {
        return /^\s*\d+\s.*/u.test(s);
    }

    newW(c) {
        const cDy = 0;
        const cMxt = 1;
        const cMnT = 2;

        const dy = parseInt(c[cDy], 10);
        const mxT = parseInt(this.substring(c[cMxt], "*"), 10);
        const mnT = parseInt(this.substring(c[cMnT], "*"), 10);

        return new W(dy, mxT, mnT);
    }

    substring(s1, s2) {
        return s1.endsWith(s2) ? s1.slice(0, -s2.length) : s1;
    }
}

module.exports = { FindW };

File: java/src/main/java/algorithm/FindW.java
package algorithm;

import java.io.IOException;
import java.util.List;

public class FindW {

    private final Helper helper = new Helper();
    private final String fn;

    public FindW(String fn) {
        this.fn = fn;
    }

    public String find(OneTwo ot) throws IOException {
        List<String> data = helper.l(fn);
        List<W> wData = p(data);
        W w = helper.find(wData, ot);
        return w.r();
    }

    private List<W> p(List<String> data) {
        return helper.p(data, //
                this::startsWithDayOfMonth, //
                this::newW);
    }

    private boolean startsWithDayOfMonth(String s) {
        return s.matches("\\s*\\d+\\s.*");
    }

    private W newW(String[] c) {
        int cDy = 0;
        int cMxt = 1;
        int cMnT = 2;

        int dy = Integer.parseInt(c[cDy]);
        int mxT = Integer.parseInt(substring(c[cMxt], "*"));
        int mnT = Integer.parseInt(substring(c[cMnT], "*"));

        return new W(dy, mxT, mnT);
    }

    private String substring(String s1, String s2) {
        if (s1.endsWith(s2)) {
            int b = s1.length() - s2.length();
            return s1.substring(0, b);
        }
        return s1;
    }

}
